src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc

```diff
@@ -83,21 +144,200 @@ PluggableAuthExternalAccountCredentials::
+    FinishRetrieveSubjectToken(
+        "", GRPC_ERROR_CREATE(
+                "Pluggable Auth executables need to be explicitly allowed to "
+                "run by setting the GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES "
+                "environment variable to 1."));
+    return;
+  }
+  struct SliceWrapper {
+    ~SliceWrapper() { CSliceUnref(slice); }
+    grpc_slice slice = grpc_empty_slice();
+  };
+  if (output_file_path_ != "") {
+    SliceWrapper content_slice;
+    grpc_error_handle error =
+        grpc_load_file(output_file_path_.c_str(), 0, &content_slice.slice);
+    if (error.ok()) {
+      absl::string_view output_file_content =
+          StringViewFromSlice(content_slice.slice);
+      CreateExecutableResponse(std::string(output_file_content));
+      if (executable_response_.success &&
+          !isExpired(executable_response_.expiration_time)) {
+        FinishRetrieveSubjectToken(executable_response_.subject_token,
+                                   absl::OkStatus());
+        return;
+      }
+    }
+  }
+  std::vector<std::string> envp_vector = {
+      absl::StrFormat("GOOGLE_EXTERNAL_ACCOUNT_AUDIENCE=%s", options.audience),
+      absl::StrFormat("GOOGLE_EXTERNAL_ACCOUNT_TOKEN_TYPE=%s",
+                      options.subject_token_type),
+      absl::StrFormat("GOOGLE_EXTERNAL_ACCOUNT_INTERACTIVE=%d", 0),
+      absl::StrFormat(
+          "GOOGLE_EXTERNAL_ACCOUNT_IMPERSONATED_EMAIL=%s",
+          get_impersonated_email(options.service_account_impersonation_url))};
+  if (output_file_path_ != "")
+    envp_vector.push_back(absl::StrFormat(
+        "GOOGLE_EXTERNAL_ACCOUNT_OUTPUT_FILE=%s", output_file_path_));
+  int environ_count = 0, i = 0;
+  while (environ[environ_count] != nullptr) environ_count += 1;
+  char** envp = (char**)gpr_malloc(sizeof(char*) *
+                                   (environ_count + envp_vector.size() + 1));
+  for (; i < environ_count; i++) envp[i] = portable_str_dup(environ[i]);
+  for (int j = 0; j < envp_vector.size(); i++, j++)
+    envp[i] = portable_str_dup(envp_vector[j].c_str());
+  envp[i] = nullptr;
+  Subprocess* subprocess = new Subprocess();
+  std::packaged_task<bool(Subprocess*, std::string, char**, std::string*,
+                          std::string*)>
+      run_executable_task(run_executable);
+  std::future<bool> executable_output_future = run_executable_task.get_future();
+  std::string output_string, error_string;
+  std::thread thr(std::move(run_executable_task), subprocess, command_, envp,
+                  &output_string, &error_string);
+  if (executable_output_future.wait_for(std::chrono::seconds(
+          executable_timeout_ms_ / 1000)) != std::future_status::timeout) {
+    thr.join();
+    if (!executable_output_future.get()) {
+      FinishRetrieveSubjectToken(
+          "", GRPC_ERROR_CREATE(absl::StrFormat(
+                  "Executable failed with error: %s", error_string)));
```

Comment:
Can we get the error code as well or just the stderror here?

Reply:
This case is when the executable crashes and doesn't produce output/output_file that are going to have error_code and error_message .